src/mgg/gmtlegs.1
.TH GMTLEGS 1 "Feb 27 2014" "GMT 4.5.13 (SVN)" "Generic Mapping Tools"
.SH NAME
gmtlegs \- Find cruises in a given region
.SH SYNOPSIS
\fBgmtlegs\fP \fB\-R\fP\fIwest\fP/\fIeast\fP/\fIsouth\fP/\fInorth\fP[\fBr\fP] [ \fB\-G\fP ] [ \fB\-M\fP ] [ \fB\-T\fP ] [ \fB\-L\fP ] [ \fB\-V\fP ]
.SH DESCRIPTION
\fBgmtlegs\fP will report the name of all the cruises that happened to be wholly or partially inside the
specified region. As an option, only those cruises that collected a certain type of data can be
reported.
.TP
\fB\-R\fP
\fIwest, east, south,\fP and \fInorth\fP specify the Region of interest, and you may specify them
in decimal degrees or in [+-]dd:mm[:ss.xxx][W|E|S|N] format.  Append \fBr\fP if lower left and upper right
map coordinates are given instead of w/e/s/n.  The two shorthands \fB\-Rg\fP and \fB\-Rd\fP stand for global domain
(0/360 and -180/+180 in longitude respectively, with -90/+90 in latitude). Alternatively, specify the name
of an existing grid file and the \fB\-R\fP settings (and grid spacing, if applicable) are copied from the grid.
.SH OPTIONS
No space between the option flag and the associated arguments
.TP
\fB\-G\fP
Report cruises that collected Gravity data. [Default is any data].
.TP
\fB\-M\fP
Report cruises that collected Magnetics. [Default is any data].
.TP
\fB\-T\fP
Report cruises that collected Topography. [Default is any data].
.TP
\fB\-L\fP
Long output, i.e., cruisenames and data types available. [Default is cruisenames only].
.TP
\fB\-V\fP
Selects verbose mode, which will send progress reports to stderr [Default runs "silently"].
.SH EXAMPLES
To find all cruises around Hawaii that collected gravity and topography:
.br
.sp
\fBgmtlegs\fP \fB\-R\fP-162/-152/18/25 \fB\-G\fP \fB\-T\fP > hawaii_gt.legs
.br
.sp
To find all cruises with magnetics:
.br
.sp
\fBgmtlegs\fP \fB\-R\fP 0/360/-90/90 \fB\-M\fP > all_m.legs
.SH "SEE ALSO"
.IR GMT (1),
.IR gmt2bin (1),
.IR binlegs (1)
.SH REFERENCES
Wessel, P., and W. H. F. Smith, 2014, The Generic Mapping Tools (GMT) version
4.5.12 Technical Reference & Cookbook, SOEST/NOAA.
.br
Wessel, P., and W. H. F. Smith, 1998, New, Improved Version of Generic Mapping
Tools Released, EOS Trans., AGU, 79 (47), p. 579.
.br
Wessel, P., and W. H. F. Smith, 1995, New Version of the Generic Mapping Tools
Released, EOS Trans., AGU, 76 (33), p. 329.
.br
Wessel, P., and W. H. F. Smith, 1995, New Version of the Generic Mapping Tools
Released, http://www.agu.org/eos_elec/95154e.html, Copyright 1995 by the
American Geophysical Union.
.br
Wessel, P., and W. H. F. Smith, 1991, Free Software Helps Map and Display Data,
EOS Trans., AGU, 72 (41), p. 441.
.br


ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 12925 2014-02-20 22:02:24Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
201x-yy-zz	pwessel
	GMT 5.1.1 finalized and released.
-----------------------------------------------------------------------------------------
2014-02-20	pwessel
	* grdcontour.c:		The -Zscale/offset option was applied incorrectly due to a
				change in the underlying scaling function.
	* grdvolume.c:		---"---
2014-02-10	pwessel
	*gmt2kml.c:		Make sure the default style IDs are unique for each process.

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.
2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		We consider the change in SDIST, PDIST to return degrees instead
				of km to be a bug as it makes it difficult to handle geodesics
				for which degrees make no sense.  Thus we have consolidated all
				the ?DIST operators to give km if geographic data are involved.
				Example 23 has been adjusted accordingly, as has several test
				scripts.  Please check your own scripts for SDIST.

2014-01-09	pwessel
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
2014-01-09	remko
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
	* grdmath.c:		LDIST and PDIST will again produce km when using geodesics.

2014-01-01	pwessel
	* x2sys/x2sys.c:	Mishandled the assignment of segment number for each record.
	* gmt_support.c:	Allow GMT_crossover to deal with global wrapped data.

2013-12-31	pwessel
	* gmt2kml.c:		Now using GMT_Put_Record instead of old stdout printing.
				This addresses issue #473.

2013-12-10	pwessel
	* grdvector.c:		Vector direction did not adjust when negative scales were
				used in -JX or -Jx.
	* psxy[z].c		---"---

2013-12-08	remko
	* psimage.c:		Now properly plots transparent images and makes -Gt work.

2013-11-26	pwessel
	* gmt_io.c:		GMT would crash on a blank line with no newline at end at all.
	* pstext.c:		pstext did not process GMT4-style input properly.

2013-11-26	jluis
	* potential/gravfft.c:	Let -E options default to 3.
				Let users select the -N+l (leave trend alone). That is, do not overwrite with default mode.
				Let the default detrend be remove MID point.
				Explicitly set -N+a in gravfft command to avoid failures due to coming changes in gravfft.
2013-11-25	pwessel
	* grdlandmask.c:	Failed to fill partial bins with no coastlines when the region
				crossed Greenwich [issue #451].

2013-11-21	pwessel
	* pscontour.c:		Let pscontour output contours to file [issue #448].
	* pscoast.c:		Prevent users from not giving proper projection [issue #449].

2013-11-19	jluis
	* mgd77/mgd77magref.c:	Needed to reset the t_col variable when constant altitude [issue #446].

2013-11-19	pwessel
	* gmt_sph.c:		Fix invalid read in ssrfpack_grid due to par array being of length 2 rather than 3.

2013-11-18	pwessel
	* gmt_grdio.c:		Place the old auto-detect of 4-byte binary float vs int under compatibility mode for backwardness.

2013-11-16	pwessel
	* grdvector.c:		Address shortcomings of geovectors when wrapped and clipped.

2013-11-15	pwessel
	* geovectors.c:		Fix geovector when shrinking is selected.

2013-11-13	pwessel
	* gmt_api.c:		Fix API issues pointed out by Ian.

2013-11-13	fwobbe
	* GMT.app:		Inherit user shell in GMT.app.

2013-11-12	pwessel
	* gmt_init.c:		grdmath could corrupt the gmt.history file [issue # 438].
 	
2013-11-10	pwessel
	* gmt_support.c:	Fix symbol-size issue in psxy -Sc [issue #435].
	
2013-11-09	pwessel
	* grdvector.c:		Did not convert direction to azimuths for geo vectors.

2013-11-08	pwessel
	* grdvector.c:		Fix grdvector -I problem [issue #434].
	* ps2raster.c:		Under Mavericks, strcpy would exit if the to and from pointers
-----------------------------------------------------------------------------------------
2013-11-05	pwessel
	GMT 5.1.0 finalized and released.
-----------------------------------------------------------------------------------------

2013-11-01	remko
	* example_10.sh:	Changed input to number of languages; multi-color columns

2013-10-31	pwessel
	* gmt_dcw.c:		Added option to specify code as =<continent> which will
				automatically be expanded to all the countries in that
				continent.  E.g., -F=AF+gred will paint Africa read.
				Fixed memory leaks.

2013-10-23	pwessel
	* gmtinfo.c:		The new name of minmax.c, the latter will be honored
				when GMT compatibility mode is 4.
	* grdmath.c:		Fixed porting bugs for CURV and D2DY2; added test ops.sh

2013-10-22	pwessel
	* minmax.c:		Added -D to adjust region found with -I
	* makecpt.c:		Better reporting of gaps or overlaps in CPTs

2013-10-05	pwessel
	* greenspline.c:	Needed -f in order to select -f0T input, plus it made assumptions
				about lon,lat despite not being selected.

2013-10-04	pwessel
	* gmt.history:		This replaces the old .gmtcommands file name.

2013-09-28	remko
	* gmt_init.c:		Moved origin of JD by one day (24 Nov instead of 25 Nov).
	* ConfigUserTemplate.cmake:	Added option GMT_DOCS_DEPEND_ON_GMT

2013-09-27	pwessel
	* share/cpt:		Added 10 new CPTs, mostly for bathymetry and topography.

2013-09-14	pwessel

src/mgg/gmtlegs.c
/*--------------------------------------------------------------------
 *	$Id: gmtlegs.c 10173 2014-01-01 09:52:34Z pwessel $
 *
 *    Copyright (c) 1991-2014 by P. Wessel and W. H. F. Smith
 *    See README file for copying and redistribution conditions.
 *--------------------------------------------------------------------*/
/*
 * gmtlegs - A program that locates all legs within a given
 * area, or only those legs that have certain types of data.
 * The program differs from LOC on VAX in that it checks each
 * 1 by 1 degree block to see if the leg actually crossed it.
 *
 * Author:	Paul Wessel
 * Date:	25-MAY-1987
 * Revised:	13-FEB-1991
 * Version 2.0: 21-Jun-1991
 * Version 3.2: 10-MAR-1999
 */

#include "gmt.h"

#define MSIZE 64000

struct INFO {
	char legname[16];
	int leg_id;
	int gmt;
	GMT_LONG mark;
	struct INFO *next_info;
};

int main (int argc, char **argv)
{
	int i, ninfo, n_found = 0, bin, leg, no, flagin, nlegs, gmt, gmtflag = 0;
	size_t not_used = 0;
	char lname[16], line[BUFSIZ], file[BUFSIZ];
	double west, east, south, north;
	FILE *fleg = NULL, *fbin = NULL;
	GMT_LONG error = FALSE, full = FALSE, verbose = FALSE, inside(double w, double e, double s, double n, int binnr);
	struct INFO *leginfo[MSIZE], *ptr[MSIZE], *make_info(char *name, int id_no, int flag);
	static char *data[8] = { "-", "G", "M", "GM", "T", "GT", "MT", "GMT" };

	argc = (int)GMT_begin (argc, argv);

	west = east = south = north = 0.0;
  	gmtflag = 0;

	for (i = 1; i < argc; i++) {
  		if (argv[i][0] == '-') {
  			switch(argv[i][1]) {
  				case 'G':
  					gmtflag += 1;
  					break;
  				case 'L':
  					full = TRUE;
  					break;
  				case 'M':
  					gmtflag += 2;
  					break;
				case 'R':
					sscanf (&argv[i][2], "%lf/%lf/%lf/%lf", &west, &east, &south, &north);
					break;
  				case 'T':
  					gmtflag += 4;
  					break;
				case 'V':
					verbose = TRUE;
					break;
  				default:
  					error = TRUE;
  					break;
  			}
  		}
  		else
  			error = TRUE;
  	}
  	while (west > east) west -= 360.0;
  	if (south >= north) error = TRUE;
  	if (west >= east) error = TRUE;
  	if (argc == 1 || error) {
  		fprintf(stderr,"usage: gmtlegs -R<west>/<east>/<south>/<north> [-G] [-L] [-M] [-T] [-V]\n");
  		fprintf(stderr,"	-R <west>, <east>, <south>, <north> is region in degrees.\n");
  		fprintf(stderr,"	-G (gravity) -M (magnetics) -T (topography). Choose any combination.\n");
  		fprintf(stderr,"	   Legs having at least the specified data are returned. [Default is any data]\n");
  		fprintf(stderr, "	-L gives a long listing (legname and datatypes). [Default is legnames only]\n");
  		fprintf(stderr, "	-V (verbose) prints number of legs found\n");
  		exit (EXIT_FAILURE);
  	}

  	GMT_getsharepath ("mgg", "gmt_legs", ".d", file);
  	if ((fleg = fopen (file, "r")) == NULL) {
		fprintf (stderr,"gmtlegs: Could not open %s\n", file);
		exit (EXIT_FAILURE);
	}
  	GMT_getsharepath ("mgg", "gmt_index", ".b", file);
	if ((fbin = fopen (file, "rb")) == NULL) {
		fprintf(stderr,"gmtlegs: Could not open %s\n", file);
		exit (EXIT_FAILURE);
	}

	/* Make sure that WESN is in whole degrees, truncate if necessary */

	west = floor (west);
	east = ceil (east);
	south = floor (south);
	north = ceil (north);

	/* Read info about each leg */

	ninfo = 0;
	while (fgets (line, BUFSIZ, fleg)) {
		sscanf (line, "%s %d %d", lname, &no, &flagin);
		ptr[no] = leginfo[ninfo++] = make_info (lname, no, flagin);
	}
	fclose (fleg);

	/* Start reading the index-file */

	while (fread ((void *)&bin, (size_t)4, (size_t)1, fbin) != 0) {
		not_used = fread ((void *)&nlegs, (size_t)4, (size_t)1, fbin);
		if (inside (west, east, south, north, bin)) {
			for (i = 0; i < nlegs; i++) {
				not_used = fread ((void *)&leg, (size_t)4, (size_t)1, fbin);
				gmt = leg & 15;
				leg >>= 4;
				if ((gmt & gmtflag) == gmtflag) ptr[leg]->mark = TRUE;
			}
		}
		else
			fseek(fbin, (long)(nlegs*4), 1);
	}
	fclose(fbin);

	for (i = 0; i < ninfo; i++) {
		if (leginfo[i]->mark) {
			(full) ? (printf("%s\t%s\n",leginfo[i]->legname,
					  data[leginfo[i]->gmt])) :
				 (printf("%s\n", leginfo[i]->legname));
			n_found++;
		}
	}
	if (verbose) fprintf (stderr, "gmtlegs: found %d legs\n", n_found);
	exit (EXIT_SUCCESS);
}

struct INFO *make_info(char *name, int id_no, int flag)
{
	struct INFO *new;
	new = (struct INFO *) GMT_memory (VNULL, (size_t)1, sizeof (struct INFO), "gmtlegs");
	strcpy (new->legname,name);
	new->leg_id = id_no;
	new->gmt = flag;
	new->mark = FALSE;
	new->next_info = 0;
	return (new);
}

GMT_LONG inside (double w, double e, double s, double n, int binnr)
{
	double lat, lon;
	lon = (double) (binnr % 360);
	while (lon > e) lon -= 360.0;
	if (lon < w || lon >= e) return (FALSE);
	lat = (double)(binnr / 360) - 90.0;
	if (lat < s || lat >= n) return (FALSE);
	return (TRUE);
}
